internal/tuning.go

```diff
@@ -0,0 +1,266 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+
+	"go.temporal.io/sdk/internal/common/metrics"
+)
+
+// WorkerTuner allows for the dynamic customization of some aspects of worker behavior.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type WorkerTuner interface {
+	GetWorkflowTaskSlotSupplier() SlotSupplier
+	GetActivityTaskSlotSupplier() SlotSupplier
+	GetLocalActivitySlotSupplier() SlotSupplier
+}
+
+// SlotPermit is a permit to use a slot.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type SlotPermit struct {
+	//lint:ignore U1000 pointless to guarantee uniqueness for now
+	int
+}
+
+// SlotReserveContext contains information that SlotSupplier instances can use during
+// reservation calls.
+type SlotReserveContext interface {
+	TaskQueue() string
+	NumIssuedSlots() int
+}
+
+// SlotSupplier controls how slots are handed out for workflow and activity tasks as well as
+// local activities when used in conjunction with a WorkerTuner.
+//
+// Currently, you cannot implement your own slot supplier. You can use the provided
+// FixedSizeSlotSupplier and ResourceBasedSlotSupplier slot suppliers.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type SlotSupplier interface {
+	// ReserveSlot is called before polling for new tasks. The implementation should block until
+	// a slot is available, then return a permit to use that slot. Implementations must be
+	// thread-safe.
+	//
+	// Any returned error besides context.Canceled will be logged and the function will be retried.
+	ReserveSlot(ctx context.Context, reserveCtx SlotReserveContext) (*SlotPermit, error)
+	// TryReserveSlot is called when attempting to reserve slots for eager workflows and activities.
+	// It should return a permit if a slot is available, and nil otherwise. Implementations must be
+	// thread-safe.
+	TryReserveSlot(reserveCtx SlotReserveContext) *SlotPermit
+
+	MarkSlotUsed()
```

Comment:
How come these APIs don't take a slot context?

Reply:
I'll add them. I was hoping I wouldn't have to fully pub the API like I did, but now that it is I'll add them, even if they're gonna be mostly empty.